lib/Controllers/base.js

```diff
@@ -106,30 +100,91 @@ Controller.prototype._control = function(req, res) {
       if (!self[hook])
         return;
 
-      var functions = Array.isArray(self[hook]) ? self[hook] : [ self[hook] ];
-      functions.forEach(function(f) {
-        work.push(function(callback) {
-          if (skip) return callback();
-          _callback = callback;
-
-          f.call(self, req, res, context);
-        });
+      hookChain = hookChain.then(function runHook(skip) {
+        if (skip) return true;
+
+        var functions = Array.isArray(self[hook]) ? self[hook] : [self[hook]];
+
+        // return the function chain. This means if the function chain resolved
+        // to skip then all the remaining hooks on this milestone will also be
+        // skipped and we will go to the next milestone
+        return functions.reduce(function(prev, current) {
+          return prev.then(function runHookFunction(skipNext) {
+
+            // if any asked to skip keep returning true to avoid calling further
+            // functions inside this hook
+            if (skipNext) return true;
+
+            var decisionPromise = new Promise(function(resolve) {
+              _.assign(context, {
+                skip: function() {
+                  resolve(context.skip);
+                },
+                stop: function() {
+                  resolve(new errors.RequestCompleted());
+                },
+                continue: function() {
+                  resolve(context.continue);
+                },
+                error: function(err) {
```

Comment:
It would be nice if this function accepted either `EpilogueError` or `status, errorMessage, [errors]`. The reasoning behind this is that in order to properly return an error in middleware that exists outside of epilogue, one would need to `require('epilogue').Errors.EpilogueError` and always throw an instance of that. I think it would be cool if we could just return parameters in an error handler and just have epilogue build up `EpilogueError` for us.